src/widgets/scrollbar.rs

```diff
@@ -438,206 +358,227 @@ impl<'a> Scrollbar<'a> {
+    ///
+    /// The scrollbar has 3 parts of note:
+    /// - `<═══█████═══════>`: full scrollbar
+    /// - ` ═══             `: track start part
+    /// - `    █████        `: thumb part part
+    /// - `         ═══════ `: track end part
+    ///
+    /// This method returns the length of the start, thunb, and end as a tuple.
+    fn part_lengths(&self, area: Rect, state: &mut ScrollbarState) -> (usize, usize, usize) {
+        let track_len = self.track_length_excluding_arrow_heads(area) as f64;
+        let viewport_len = self.viewport_length(area) as f64;
+
+        let content_length = state.content_length as f64;
+        // Clamp the position to show at least one line of the content, even if the content is
+        let position = state.position.min(state.content_length - 1) as f64;
+
+        // vscode style scrolling behavior (allow scrolling past end of content)
+        let scrollable_content_len = content_length + viewport_len - 1.0;
+        let thumb_start = position * track_len / scrollable_content_len;
+        let thumb_end = (position + viewport_len) * track_len / scrollable_content_len;
+
+        // We round just the positions (instead of floor / ceil), and then calculate the sizes from
+        // those positions. Rounding the sizes instead causes subtle off by 1 errors.
+        let track_start_len = thumb_start.round() as usize;
+        let thumb_end = thumb_end.round() as usize;
+
+        let thumb_len = thumb_end.saturating_sub(track_start_len);
+        let track_end_len = track_len as usize - track_start_len - thumb_len;
+
+        (track_start_len, thumb_len, track_end_len)
+    }
+
+    fn scollbar_area(&self, area: Rect) -> Rect {
+        match self.orientation {
+            ScrollbarOrientation::VerticalLeft => area.columns().next(),
+            ScrollbarOrientation::VerticalRight => area.columns().last(),
+            ScrollbarOrientation::HorizontalTop => area.rows().next(),
+            ScrollbarOrientation::HorizontalBottom => area.rows().last(),
+        }
+        .expect("Scrollbar area is empty") // this should never happen as we check for empty area
+    }
+
+    /// Calculates length of the track excluding the arrow heads
+    ///
+    /// ```plain
+    ///        ┌────────── track_length
+    ///  vvvvvvvvvvvvvvv
+    /// <═══█████═══════>
+    /// ```
+    fn track_length_excluding_arrow_heads(&self, area: Rect) -> u16 {
+        let start_len = self.begin_symbol.map(|s| s.width() as u16).unwrap_or(0);
+        let end_len = self.end_symbol.map(|s| s.width() as u16).unwrap_or(0);
+        let arrows_len = start_len + end_len;
+        if self.orientation.is_vertical() {
+            area.height.saturating_sub(arrows_len)
+        } else {
+            area.width.saturating_sub(arrows_len)
+        }
+    }
+
+    fn viewport_length(&self, area: Rect) -> u16 {
+        if self.orientation.is_vertical() {
+            area.height
+        } else {
+            area.width
```

Comment:
This should probably use `viewport_content_length`, right?

Reply:
I'm not sure that I'm clear enough by what the original intent of the viewport content length was enough to make the change.
The example on ScrollbarState confuses me as it leaves several ambiguous interpretations:
- Is 2 the number of items that are visible in the viewport, or the size of each item?
- what happens if the multi-line items are not a fixed size?

Reply:
I think the original intent was "viewport size" but it was poorly named and documented by me to mean number of items in the viewport. 

For example, a viewport of size 2 where each item is 1 line in height should behave the same as a "2 items in the viewport". 

I didn't map those concepts on to each other though. The implementation under the hood does exactly that though.

Reply:
ok - I'm going to add a viewport_size method, make the underlying Option<usize> instead of 0 and use it

Reply:
That works!

Reply:
actually ended up not doing this. Option breaks serialization, and we need to discuss this further.